git/ignore.go

```diff
@@ -0,0 +1,275 @@
+}
+
+func (im IgnoreMatch) LineString() string {
+	lineNum := ""
+	if im.LineNum != 0 {
+		lineNum = strconv.Itoa(im.LineNum)
+	}
+	return lineNum
+}
+
+func (ip IgnorePattern) Matches(ignorePath string, isDir bool) bool {
+	if !strings.HasPrefix(ignorePath, ip.Scope.String()) {
+		return false
+	}
+
+	rel, err := filepath.Rel("/"+ip.Scope.String(), "/"+ignorePath)
+
+	// Not in scope of this pattern
+	if err != nil || strings.HasPrefix(rel, ".") {
+		return false
+	}
+
+	return matchesGlob("/"+rel, isDir, ip.Pattern)
+}
+
+// Returns the standard ignores (.gitignore files and various global sources) that
+//  should be used to determine whether the provided files are ignored. These patterns
+//  can be used with these files with ApplyIgnore() to find whether each file is ignored
+//  or not.
+func StandardIgnorePatterns(c *Client, paths []File) ([]IgnorePattern, error) {
+	alreadyParsed := make(map[string]bool)
+	finalPatterns := []IgnorePattern{}
+
+	for _, path := range paths {
+		abs, err := filepath.Abs(path.String())
+		if err != nil {
+			return nil, err
+		}
+
+		// Let's check that this path is in the git work dir first
+		wdpath, err := filepath.Rel(c.WorkDir.String(), abs)
+		if err != nil || strings.HasPrefix(wdpath, "..") {
+			return nil, fmt.Errorf("Path %v is not in the git work directory.", path.String())
+		}
+
+		// this could be the work dir specified as '.' so let's start with that.
+		dir := filepath.Dir(abs)
+
+		for {
+			gitignore := filepath.Join(dir, ".gitignore")
+			if _, ok := alreadyParsed[gitignore]; ok {
+				if dir == c.WorkDir.String() {
+					break
+				}
+				dir = filepath.Dir(dir)
+				continue
+			}
+			alreadyParsed[gitignore] = true
+
+			scope, err := filepath.Rel(c.WorkDir.String(), dir)
+			if err != nil {
+				return []IgnorePattern{}, err
+			}
+
+			if scope == "." {
+				scope = ""
+			}
+
+			ignorePatterns, err := ParseIgnorePatterns(c, File(gitignore), File(scope))
+			if err != nil {
+				return []IgnorePattern{}, err
+			}
+
+			finalPatterns = append(finalPatterns, ignorePatterns...)
+
+			if dir == c.WorkDir.String() || dir == filepath.Dir(c.WorkDir.String()) {
+				break
+			}
+			dir = filepath.Dir(dir)
+		}
+	}
+
+	ignorePatterns, err := ParseIgnorePatterns(c, File(filepath.Join(c.GitDir.String(), "info/exclude")), File(""))
+	if err != nil {
+		return []IgnorePattern{}, err
+	}
+
+	finalPatterns = append(finalPatterns, ignorePatterns...)
+
+	return finalPatterns, err
+}
+
+func ParseIgnorePatterns(c *Client, patternFile File, scope File) ([]IgnorePattern, error) {
+	log.Printf("Parsing %s for ignore patterns\n", patternFile)
+	_, err := patternFile.Lstat()
+	if os.IsNotExist(err) {
+		return []IgnorePattern{}, nil
+	}
+
+	if err != nil {
+		return []IgnorePattern{}, err
+	}
+
+	file, err := patternFile.Open()
+	if err != nil {
+		return []IgnorePattern{}, err
+	}
+	defer file.Close()
+
+	ignorePatterns := []IgnorePattern{}
+	reader := bufio.NewReader(file)
```

Comment:
I haven't looked into this loop in detail, but from a cursory glance it seems that this might be easier using a bufio.Scanner to scan lines instead of a bufio.Reader

Reply:
I wish that were the case, but unfortunately buffo.Scanner will miss the last line of a stream if there is no newline character at the end. :( This is why there is the pain of having to use the reader directly.